src/Ryujinx.Graphics.Gpu/Engine/Dma/DmaClass.cs

```diff
@@ -279,7 +279,11 @@ private void DmaCopy(int argument)
                 bool completeSource = IsTextureCopyComplete(src, srcLinear, srcBpp, srcStride, xCount, yCount);
                 bool completeDest = IsTextureCopyComplete(dst, dstLinear, dstBpp, dstStride, xCount, yCount);
 
-                if (completeSource && completeDest)
+                // Try to set the texture data directly,
+                // but only if we are doing a complete copy,
+                // and not for block linear to linear copies, since those are typically accessed from the CPU.
+
+                if (completeSource && completeDest && (srcLinear || !dstLinear))
```

Comment:
```suggestion
                if (completeSource && completeDest && !(dstLinear && !srcLinear))
```

This condition really confused me when I first saw it, which I thought was odd given the explanation of the PR which should have made it obvious.

I think the reason for that is that it uses an OR to exclude a single case using all others, rather than using an AND to identify the single case and negating it. You have to think about all the cases that the condition is true, rather than just the case where it isn't.

It does add an extra negation, but I think it's clearer.

Reply:
For me, everything with an additional negation makes things more confusing, because I have to think "actually, this case is *not* allowed, so the allowed cases are all others except this". But I don't mind changing it.